cmd/cainjector/app/start.go

```diff
@@ -187,58 +197,45 @@ func (o InjectorControllerOptions) RunInjectorController(ctx context.Context) er
 		})
 	}
 
-	g.Go(func() (err error) {
-		defer func() {
-			o.log.Error(err, "manager goroutine exited")
-		}()
-
-		if err = mgr.Start(gctx); err != nil {
-			return fmt.Errorf("error running manager: %v", err)
+	// If cainjector has been configured to watch Certificate CRDs
+	// (--watch-certificates=true), poll kubeapiserver for 5 minutes or till
+	// certificate CRD is found.
+	if o.WatchCerts {
+		directClient, err := client.New(mgr.GetConfig(), client.Options{
+			Scheme: mgr.GetScheme(),
+			Mapper: mgr.GetRESTMapper(),
+		})
+		if err != nil {
+			return fmt.Errorf("failed to create client: %w", err)
 		}
-		return nil
-	})
-
-	select {
-	case <-gctx.Done(): // Exit early if we are shutting down or if the manager has exited with an error
-		// Wait for error group to complete and return
-		return g.Wait()
-	case <-mgr.Elected(): // Don't launch the controllers unless we have been elected leader
-		// Continue with setting up controller
-	}
-
-	// Retry the start up of the certificate based controller in case the
-	// cert-manager CRDs have not been installed yet or in case the CRD API is
-	// not working. E.g. The conversion webhook has not yet had its CA bundle
-	// injected by the secret based controller, which is launched in its own
-	// goroutine.
-	// When shutting down, return the last error if there is one.
-	// Never retry if the controller exits cleanly.
-	g.Go(func() (err error) {
-		for {
-			err = cainjector.RegisterCertificateBased(gctx, mgr, o.Namespace)
-			if err == nil {
-				return
+		err = wait.PollImmediate(time.Second, time.Minute*5, func() (bool, error) {
+			certsCRDName := types.NamespacedName{Name: "certificates.cert-manager.io"}
+			certsCRD := apiext.CustomResourceDefinition{}
+			err := directClient.Get(ctx, certsCRDName, &certsCRD)
```

Comment:
If we didn't have the RBAC to watch CRDs, I would have used the discovery API instead. Not sure if there is a possible difference between what CRDs are installed and what the discovery API returns.

Reply:
I was considering that alternative- my thinking was that if we already know exactly which CRD we want then sending one GET request for that makes more sense than using discovery API to list and parse all